src/conf_mode/firewall.py

```diff
@@ -192,6 +192,36 @@ def verify_rule(firewall, rule_conf, ipv6):
         if not {'count', 'time'} <= set(rule_conf['recent']):
             raise ConfigError('Recent "count" and "time" values must be defined')
 
+    if 'gre' in rule_conf:
+        if dict_search_args(rule_conf, 'protocol') != 'gre':
+            raise ConfigError('Protocol must be gre when matching GRE flags and fields')
+
+        gre_checking_key = dict_search_args(rule_conf, 'gre', 'key')
+        if gre_checking_key:
+            if dict_search_args(rule_conf, 'gre', 'version') == 'pptp':
+                raise ConfigError('GRE tunnel keys are not present in PPTP')
+
+            if dict_search_args(rule_conf, 'gre', 'flags', 'checksum') is None and \
+               dict_search_args(rule_conf, 'gre', 'flags', 'not', 'checksum') is None:
+                # There is no builtin match in nftables for the GRE key, so we need to do a raw lookup.
+                # The offset of the key within the packet shifts depending on the C-flag. 
+                # 99% of the time, nobody will have checksums enabled - it's usually a manual config option. 
+                # We can either assume it is unset unless otherwise directed 
+                # (confusing, requires doco to explain why it doesn't work sometimes)
+                # or, demand an explicit selection to be made for this specific match rule. 
+                # This check enforces the latter. The user is free to create rules for both cases. 
+                raise ConfigError('Matching GRE tunnel key requires an explicit checksum flag match. For most cases, use "gre flags not checksum"')
+
+        gre_flags = dict_search_args(rule_conf, 'gre', 'flags')
+        if gre_flags:
+            gre_not_flags = dict_search_args(rule_conf, 'gre', 'flags', 'not')
+            if gre_not_flags:
+                duplicates = [flag for flag in gre_flags if flag in gre_not_flags]
+                if duplicates:
+                    raise ConfigError(f'Cannot match a GRE flag as set and not set')
+                if 'key' in gre_not_flags and gre_checking_key:
+                    raise ConfigError('Matching GRE tunnel key implies "flags key", cannot specify "flags not key"')
```

Comment:
Here mutually exclusive options are inevitable, but I still think these commands might be better as `set ... rule 1 gre flags key value <...>` and `set ... rule 1 gre flags key unset`.